objects/selection_acquire.c

```diff
@@ -0,0 +1,240 @@
+    xcb_window_t window;
+    /** Timestamp used for acquiring the selection. */
+    xcb_timestamp_t timestamp;
+} selection_acquire_t;
+
+static lua_class_t selection_acquire_class;
+LUA_OBJECT_FUNCS(selection_acquire_class, selection_acquire_t, selection_acquire)
+
+static void
+luaA_pushatom(lua_State *L, xcb_atom_t atom)
+{
+    lua_pushnumber(L, atom);
+}
+
+static int
+selection_acquire_find_by_window(lua_State *L, xcb_window_t window)
+{
+    /* Iterate over all active selection acquire objects */
+    lua_pushliteral(L, REGISTRY_ACQUIRE_TABLE_INDEX);
+    lua_rawget(L, LUA_REGISTRYINDEX);
+    lua_pushnil(L);
+    while (lua_next(L, -2) != 0) {
+        if (lua_type(L, -1) == LUA_TUSERDATA) {
+            selection_acquire_t *selection = lua_touserdata(L, -1);
+            if (selection->window == window)
+            {
+                /* Remove table and key */
+                lua_remove(L, -2);
+                lua_remove(L, -2);
+                return 1;
+            }
+        }
+        /* Remove the value, leaving only the key */
+        lua_pop(L, 1);
+    }
+    /* Remove the table */
+    lua_pop(L, 1);
+
+    return 0;
+}
+
+static void
+selection_release(lua_State *L, int ud)
+{
+    selection_acquire_t *selection = luaA_checkudata(L, ud, &selection_acquire_class);
+
+    luaA_object_emit_signal(L, ud, "release", 0);
+
+    /* Destroy the window, this also releases the selection in X11 */
+    xcb_destroy_window(globalconf.connection, selection->window);
+    selection->window = XCB_NONE;
+
+    /* Unreference the object, it's now dead */
+    lua_pushliteral(L, REGISTRY_ACQUIRE_TABLE_INDEX);
+    lua_rawget(L, LUA_REGISTRYINDEX);
+    luaA_pushatom(L, selection->selection);
+    lua_pushnil(L);
+    lua_rawset(L, -3);
+
+    selection->selection = XCB_NONE;
+
+    lua_pop(L, 1);
+}
+
+void
+selection_handle_selectionclear(xcb_selection_clear_event_t *ev)
+{
+    lua_State *L = globalconf_get_lua_State();
+
+    if (selection_acquire_find_by_window(L, ev->owner) == 0)
+        return;
+
+    selection_release(L, -1);
+    lua_pop(L, 1);
+}
+
+void
+selection_handle_selectionrequest(xcb_selection_request_event_t *ev)
+{
+    lua_State *L = globalconf_get_lua_State();
+
+    if (ev->property == XCB_NONE)
+        /* Obsolete client */
+        ev->property = ev->target;
+
+    if (selection_acquire_find_by_window(L, ev->owner) == 0)
+    {
+        selection_transfer_reject(ev->requestor, ev->selection, ev->target, ev->time);
+        return;
+    }
+
+    selection_transfer_begin(L, -1, ev->requestor, ev->selection, ev->target,
+            ev->property, ev->time);
+
+    lua_pop(L, 1);
+}
+
+static int
+luaA_selection_acquire_new(lua_State *L)
+{
+    size_t name_length;
+    const char *name;
+    xcb_intern_atom_reply_t *reply;
+    xcb_get_selection_owner_reply_t *selection_reply;
+    xcb_atom_t name_atom;
+    selection_acquire_t *selection;
+
+    name = luaL_checklstring(L, 2, &name_length);
```

Comment:
For consistency and futureproofness considerations, can we have an `args` table even if there's (currently) only a single argument?